File: objectiveai-sdk-py/objectiveai_sdk/push_utils.py
"""Shared helpers for push (streaming chunk accumulation)."""

from __future__ import annotations


def push_by_index(self_list: list, other_list: list) -> None:
    """Merge *other_list* into *self_list* by ``index`` field.

    Items with a matching index are merged via ``push()``.
    New indices are appended.
    """
    index_map: dict[int, int] = {}
    for pos, item in enumerate(self_list):
        idx = _get_index(item)
        if idx is not None:
            index_map[idx] = pos

    for other_item in other_list:
        idx = _get_index(other_item)
        if idx is not None and idx in index_map:
            self_list[index_map[idx]].push(other_item)
        else:
            self_list.append(other_item)
            if idx is not None:
                index_map[idx] = len(self_list) - 1


def push_option(obj: object, attr: str, other_val) -> None:
    """Conditionally merge an optional sub-object field.

    Both present → delegate to push(). Only other → adopt.
    Only self / neither → no assignment (field stays unset).
    """
    self_val = getattr(obj, attr)
    if self_val is not None and other_val is not None:
        self_val.push(other_val)
    elif other_val is not None:
        setattr(obj, attr, other_val)


def push_replace(obj: object, attr: str, other_val) -> None:
    """Replace field only if other is not None (latest wins)."""
    if other_val is not None:
        setattr(obj, attr, other_val)


def push_option_int(obj: object, attr: str, other_val: int | None) -> None:
    """Sum two optional ints. Only assigns if other is not None."""
    if other_val is None:
        return
    self_val = getattr(obj, attr)
    if self_val is not None:
        setattr(obj, attr, self_val + other_val)
    else:
        setattr(obj, attr, other_val)


def push_option_string(obj: object, attr: str, other_val: str | None) -> None:
    """Concatenate two optional strings. Only assigns if other is not None."""
    if other_val is None:
        return
    self_val = getattr(obj, attr)
    if self_val is not None:
        setattr(obj, attr, self_val + other_val)
    else:
        setattr(obj, attr, other_val)


def push_option_decimal(obj: object, attr: str, other_val) -> None:
    """Sum two optional decimals/floats. Only assigns if other is not None."""
    if other_val is None:
        return
    self_val = getattr(obj, attr)
    if self_val is not None:
        setattr(obj, attr, self_val + other_val)
    else:
        setattr(obj, attr, other_val)


def push_lazy_set_true(obj: object, attr: str, other_val: bool | None) -> None:
    """Set to True only if other is True. Never overwrites to False."""
    if other_val is True:
        setattr(obj, attr, True)


def _get_index(item):
    """Extract an integer index from a model (BaseModel or RootModel)."""
    from pydantic import RootModel
    if isinstance(item, RootModel):
        inner = item.root
        if isinstance(inner, RootModel):
            inner = inner.root
        return getattr(inner, "index", None)
    return getattr(item, "index", None)

File: objectiveai-sdk-go/push_utils.go
package objectiveai

// pushOptionUint64 adds other to self. If self is nil, adopts other's value.
func pushOptionUint64(self *uint64, other *uint64) *uint64 {
	if self != nil && other != nil {
		sum := *self + *other
		return &sum
	}
	if other != nil {
		v := *other
		return &v
	}
	return self
}

// pushOptionFloat64 adds other to self. If self is nil, adopts other's value.
func pushOptionFloat64(self *float64, other *float64) *float64 {
	if self != nil && other != nil {
		sum := *self + *other
		return &sum
	}
	if other != nil {
		v := *other
		return &v
	}
	return self
}

// pushOptionString concatenates other onto self. If self is nil, adopts other.
func pushOptionString(self *string, other *string) *string {
	if self != nil && other != nil {
		s := *self + *other
		return &s
	}
	if other != nil {
		s := *other
		return &s
	}
	return self
}

// pushByIndex merges items from other into self by matching an index field.
// The getIndex function extracts the index, and push merges matching items.
func pushByIndex[T any](self *[]T, other []T, getIndex func(*T) uint64, push func(*T, *T)) {
	indexMap := make(map[uint64]int)
	for i := range *self {
		idx := getIndex(&(*self)[i])
		indexMap[idx] = i
	}
	for i := range other {
		idx := getIndex(&other[i])
		if pos, ok := indexMap[idx]; ok {
			push(&(*self)[pos], &other[i])
		} else {
			*self = append(*self, other[i])
			indexMap[idx] = len(*self) - 1
		}
	}
}

// pushByNullableIndex is like pushByIndex but the index can be nil.
// Items with nil index are always appended (never matched).
func pushByNullableIndex[T any](self *[]T, other []T, getIndex func(*T) *uint64, push func(*T, *T)) {
	indexMap := make(map[uint64]int)
	for i := range *self {
		if idx := getIndex(&(*self)[i]); idx != nil {
			indexMap[*idx] = i
		}
	}
	for i := range other {
		idx := getIndex(&other[i])
		if idx != nil {
			if pos, ok := indexMap[*idx]; ok {
				push(&(*self)[pos], &other[i])
			} else {
				*self = append(*self, other[i])
				indexMap[*idx] = len(*self) - 1
			}
		} else {
			*self = append(*self, other[i])
		}
	}
}

// pushLazySetTrue only sets to true, never overwrites to false.
func pushLazySetTrue(self *bool, other *bool) *bool {
	if other != nil && *other {
		v := true
		return &v
	}
	return self
}

// pushReplace replaces self with other if other is not nil (latest wins).
func pushReplace[T any](self *T, other *T) *T {
	if other != nil {
		return other
	}
	return self
}
